### PR TITLE
Session Delayed Para Changes / Actions Queue

### PR DESCRIPTION
**Pull request by shawntabrizi**
Fixes: https://github.com/paritytech/polkadot/issues/2300

Design Notes: https://hackmd.io/XDkKzaYMSYumhNWzRMEvCQ

This PR introduces a full session delay to any changes to parachains and parathreads in the para module and any changes to the parachains runtime configuration.

Changes:

* Shared Module for tracking the current session index and safe session to schedule changes.
* Remove multiple different queues in favor of a single `ActionQueue` which tracks all paras that need to be updated at the start of a specific session.
* `ActionQueue` is a map from `SessionIndex -> Vec<ParaId>` so we can schedule the change of a para for a future session.
* Actions are queued for `current_session + 2` session index in the future, solving the problem in issue 2300.
* Cannot queue any other changes to a para when there is already a change in the queue. As such, any para scheduling now returns a result to handle the case where we are not allowed to schedule some task.
* Configurations are queued for future session.

### CODE REVIEW

**Comment by shawntabrizi** [comment]
@rphmeier and @pepyakin 

What should be considered for `is_valid_para` given the delayed state transitions?

For example, if a para is being offboarded, should they immediately not be considered a valid para? How about in the context of HRMP channel registration?

**Comment by pepyakin** [comment]
`is_valid_para` is used exclusively in HRMP. The primary reason why it is there has to do with pruning of stale items such as channels.

So channels are opened on the nearest session boundary. If during that moment it turns out that the channel is offboarded (and HRMP's session change handler gets control after Paras') the request is declined. If this check wasn't there then the channel will be left as junk, because there will be no further signals that the para is offboarded. 

The second place where we check that is before accepting a channel open request, which just makes sure that we don't open a channel to a non-existent para, but it is kind of optional because the recipient may request offboarding anyway later in which case the first check will handle this.

I think as long as we get `is_valid_para(P) = true` for the cases when we can expect a further notification of removal is fine.

But now I am more worried about another aspect. Right now, we foresee a possibility that parachains and parathread may differ in configuration. For instance, a parachain can have N channels and a parathread can have only M channels, where N > M. For HRMP it doesn't matter that much in practice since we limit them as much as possible and do not make a difference between parachains and parathreads, but for XCMP that may become a problem.

**Comment by rphmeier** [comment]
> For instance, a parachain can have N channels and a parathread can have only M channels, where N > M. For HRMP it doesn't matter that much in practice since we limit them as much as possible and do not make a difference between parachains and parathreads, but for XCMP that may become a problem.

I think the `Initializer` should notify modules of paras deregistering, upgrading, downgrading, etc. and we can handle it in those modules. We could just cut off all channels when swapping for instance, or just cut off `N - M` when downgrading

**Comment by rphmeier** [comment]
This should not close #2300 unless it does something similar for `Configuration` updates. (disclaimer: haven't looked)

**Comment by shawntabrizi** [comment]
I am happy to update for `PendingConfig` too.

Probably best to do it in a follow up PR unless you want me to make this PR larger.

**Comment by shawntabrizi** [comment]
@rphmeier should be good to go and completely close #2300

**Comment by shawntabrizi** [comment]
One issue with this PR as it currently stands is that it is possible that a parachain cannot be queued to be offboarded. It is possible to imagine that some combinations of states could be used to grief the parachains system and prevent an offboarding.

That being said, we will always return the result of trying to schedule offboarding, so it can always be handled, but maybe we need to create a more low level API to make sure that a parachain will be offboarded in the future.

File: runtime/parachains/src/session_info.rs [hunk 1]
```diff
@@ -218,24 +225,14 @@ mod tests {
 	}
 
 	fn session_changes(n: BlockNumber) -> Option<SessionChangeNotification<BlockNumber>> {
-		match n {
-			100 => Some(SessionChangeNotification {
-				session_index: 10,
+		println!("block {:?}", n);
```
Comment (rphmeier) [inline runtime/parachains/src/session_info.rs:228]: stray println

**Review by rphmeier — APPROVED** [review]
Looks good to me! Thanks for the cleanup of tests and addition of `Shared` module

File: runtime/parachains/src/lib.rs [hunk 2]
```diff
@@ -49,11 +50,11 @@ pub use paras::ParaLifecycle;
 pub fn schedule_para_initialize<T: paras::Config>(
 	id: primitives::v1::Id,
 	genesis: paras::ParaGenesisArgs,
-) {
-	<paras::Module<T>>::schedule_para_initialize(id, genesis);
+) -> Result<(), ()> {
+	<paras::Module<T>>::schedule_para_initialize(id, genesis).map_err(|_| ())
```
Comment (rphmeier) [inline runtime/parachains/src/lib.rs:54]: An API nit. When you want to do swapping I assume you'd call `onboard()` on one and `offboard()` on the other. The first might succeed and the second might fail. You could either have some kind of atomic updating API where you update multiple paras at once or you just expose the `can_*` functions here. Something for a follow-up?

Comment (shawntabrizi) [reply runtime/parachains/src/lib.rs:54]: ack, will follow up in another PR and with more consideration specifically with the swap integration in registrar

File: runtime/common/src/paras_registrar.rs [hunk 3]
```diff
@@ -309,7 +308,7 @@ mod tests {
 	parameter_types! {
 		pub const BlockHashCount: u32 = 250;
 		pub BlockWeights: limits::BlockWeights =
-			limits::BlockWeights::with_sensible_defaults(4 * 1024 * 1024, NORMAL_RATIO);
+			frame_system::limits::BlockWeights::simple_max(1024);
```
Comment (ggwpez) [inline runtime/common/src/paras_registrar.rs:311]: Just wondering: what was the motivation here?

**Pull request merged**